Feature/insights contact merge

espl-ishita opened the pull request:
Related work: https://jira.makanaplatform.com/browse/INSIGHTS-713

# Critical Changes

# Changes

Updates to Contact Merge functionality are done to display two buttons on the Contact Merge page. First button 'Search By Contacts' will redirect to the existing contact merge functionality. The second button 'Show Duplicate Record Sets' will redirect to list of DRSs related to contact duplicate rules. This second page includes pagination to handle large number of DRS records. The 'Duplicate Record Set Name', 'Duplicate Rule' and 'First Contact Name' columns of the DRS list contains the links to the detail page of the respective records

# Issues Closed

# Community Ideas Delivered

# New Metadata

# Deleted Metadata

espl-ishita commented:
@ptewson-sfdo  The final updates to the DRS contact merge functionality are done and ready to review

ptewson-sfdo commented:
@espl-ishita - I'm just starting formal review on this, but I noticed right away that the automated builds are failing (message below). Does your team have fixes in the works? 

Could not process MDAPI response: Update of ApexPage DRS_ContactMerge: Error: Required field is missing: label

Update of ApexClass CON_ContactMerge_CTRL: Error on line 834, col 86: External string does not exist: conMergeErrorNoDRSFound

Update of ApexClass CON_ContactMerge_TEST: Error on line 834, col 86: Dependent class is invalid and needs recompilation:
Class CON_ContactMerge_CTRL : External string does not exist: conMergeErrorNoDRSFound

Update of ApexClass CON_ContactMerge_TEST2: Error on line 834, col 86: Dependent class is invalid and needs recompilation:
Class CON_ContactMerge_TEST : Dependent class is invalid and needs recompilation:
Class CON_ContactMerge_CTRL : External string does not exist: conMergeErrorNoDRSFound

Update of ApexPage CON_ContactMerge: Error: Apex class 'CON_ContactMerge_CTRL' does not exist

ptewson-sfdo commented on src/classes/DRS_ContactMerge_CTRL:
```diff
@@ -0,0 +1,1008 @@
+
+        // Add second row to select the master record. Select the one which is last modified
+        temp = new FieldRow(MASTER_KEY, 'Master Record', true, null);
+        {
```
Can remove this set of curly braces (line #771 and 781). I don't *think* they are doing anything

espl-ishita replied:
Updated

espl-ishita commented on src/classes/DRS_ContactMerge_CTRL:
```diff
@@ -0,0 +1,1008 @@
+public with sharing class DRS_ContactMerge_CTRL {
```
Updated the class name to  DRS_ContactMerge_CTRL.cls 
Added standard comment block class header (with 2020 copyright date)
Uploaded the related meta.xml file

espl-ishita commented on src/classes/CON_ContactMerge_CTRL.cls:
```diff
@@ -795,4 +821,38 @@ public with sharing class CON_ContactMerge_CTRL {
+            showContactSearch = false;
+            if (!hasDSRObjectReadPermission()) {
+                showDRSButton = false;
+                ApexPages.addMessage(new ApexPages.Message(ApexPages.Severity.ERROR, System.Label.conMergeErrorNoDRSFound));
```
Added the label conMergeErrorNoDRSFound to the CustomLabel.label and package.xml files

espl-ishita commented on DRS_ContactMerge_Test.cls:
```diff
@@ -0,0 +1,321 @@
+* @group ContactMerge
+* @group-content ../../ApexDocContent/ContactMerge.htm
+*/
+@isTest
```
Updated the test class with "IsParallel=true"

espl-ishita commented on DRS_ContactMerge_Test.cls:
```diff
@@ -0,0 +1,321 @@
+/*
+    Copyright (c) 2009, Salesforce.org
```
Updated the year in the comment block

espl-ishita commented on src/pages/CON_ContactMerge.page:
```diff
@@ -15,15 +15,29 @@
+            <div class="slds-m-top_medium slds-m-bottom_large slds-p-horizontal_large">
+                <apex:commandButton id="srchByConBtn"
+                                    action="{!showContactSearchPage}" 
+                                    value="{!$Label.conMergeSearchByContact}" 
```
Added the label conMergeSearchByContact to the CustomLabel.label and package.xml files

espl-ishita commented on src/pages/CON_ContactMerge.page:
```diff
@@ -42,228 +56,191 @@
             }
         </style>
-
+        
```
Updated

espl-ishita commented on src/pages/CON_ContactMerge.page:
```diff
@@ -15,15 +15,29 @@
+            <div class="slds-m-top_medium slds-m-bottom_large slds-p-horizontal_large">                
+                <apex:commandButton id="ShowDRS"
+                                    action="{!redirectToShowDRSList}" 
+                                    value="Show Duplicate Record Sets" 
```
Added the text "Show Duplicate Record Sets" to a custom label and added that custom label to the CustomLabel.label and package.xml files

espl-ishita commented on src/pages/DRS_ContactMerge.page:
```diff
@@ -0,0 +1,437 @@
+                <div class="slds-m-top_medium slds-m-bottom_large slds-p-horizontal_large">
+                    <apex:commandButton id="srchByConBtn"
+                                        action="{!searchByContact}" 
+                                        value="{!$Label.conMergeSearchByContact}" 
```
Added the label conMergeSearchByContact to the CustomLabel.label and package.xml files

espl-ishita commented on DRS_ContactMerge_Test.cls:
```diff
@@ -0,0 +1,321 @@
+/*
```
Uploaded the meta.xml file for this class

espl-ishita commented on DRS_ContactMerge_Test.cls:
```diff
@@ -0,0 +1,321 @@
+/*
+    Copyright (c) 2009, Salesforce.org
```
Updated

espl-ishita commented on DRS_ContactMerge_Test.cls:
```diff
@@ -0,0 +1,321 @@
+*/
+/**
+* @author Salesforce.org
+* @date 2011 (2.x)
```
Deleted this class as it was created in wrong folder and created a similar new class under src/classes .
Updated the date in that class

espl-ishita commented on src/classes/CON_ContactMerge_CTRL.cls:
```diff
@@ -184,14 +203,21 @@ public with sharing class CON_ContactMerge_CTRL {
     * @param controller The default list controller for contact. It allows us to do pagination on the page.
     */
     public CON_ContactMerge_CTRL(ApexPages.StandardSetController controller){
+        System.debug('Contact Merge page contructor first line----->');
```
Removed the debug

espl-ishita commented on src/classes/CON_ContactMerge_CTRL.cls:
```diff
@@ -795,4 +821,38 @@ public with sharing class CON_ContactMerge_CTRL {
+            showContactSearch = false;
+            if (!hasDSRObjectReadPermission()) {
+                showDRSButton = false;
+                ApexPages.addMessage(new ApexPages.Message(ApexPages.Severity.ERROR, System.Label.conMergeErrorNoDRSFound));
```
Added the label to the CustomLabel.label and package.xml files

espl-ishita commented on src/classes/DRS_ContactMerge_CTRL:
```diff
@@ -0,0 +1,1008 @@
+        // Finally completing the query by appending the table name and the filter clause
+        query += ' from Contact where id IN :mergeCandidates';
+
+        //UTIL_Debug.debug('The contact query is: ' + query);
```
Updated

espl-ishita commented on src/classes/DRS_ContactMerge_CTRL:
```diff
@@ -0,0 +1,1008 @@
+    private Boolean isNPSPReadOnly(String strField) {
+        // we have to maintain our own set of readonly fields, because
+        // RLLP_OppRollup_UTIL.ContactFields doesn't include them all!
+        List<String> listStrField = new List<String>{
```
Declared the listStrField variable at class level and static.
@ptewson-sfdo  This code section is replicated from the existing contact merge functionality i.e from CON_ContactMerge_CTRL class. Do you want us to do the same updates to CON_ContactMerge_CTRL class as well?

espl-ishita commented on src/classes/DRS_ContactMerge_CTRL:
```diff
@@ -0,0 +1,1008 @@
+
+        // Add the first row of type header with contact names as titles
+        FieldRow temp = new FieldRow('', '', false, 'header');
+        for (Sobject c: contacts) {
```
Updated the 'sObject' keyword to Contact.
@ptewson-sfdo This code section is replicated from the existing contact merge functionality i.e from CON_ContactMerge_CTRL class. Do you want us to do the same updates to CON_ContactMerge_CTRL class as well?

espl-ishita commented:
@ptewson-sfdo   I have updated the related classes and pages as per your comments. I have uploaded the respective meta.xml files. Custom labels have been added to the CustomLabels.label and package.xml files.

ptewson-sfdo commented:
@espl-ishita It looks like the automated builds are still failing - "Update of ApexPage DRS_ContactMerge: Error: Required field is missing: label". Do you have visibility to those checks and the log files that display the errors? 

![image](https://user-images.githubusercontent.com/45107660/77105984-ac700c00-69db-11ea-8a7f-89421992cf64.png)

ptewson-sfdo commented:
> @ptewson-sfdo The final updates to the DRS contact merge functionality are done and ready to review

Just FYI these updates aren't final - once I can functionally test this out there will be additional review, but first we have to get this in a state where it will deploy and the automated builds pass!

espl-ishita commented:
@ptewson-sfdo I have uploaded 2 more files inorder to solve the automatic build issues. Please have a look

ptewson-sfdo commented on src/classes/DRS_ContactMerge_CTRL.cls:
```diff
@@ -0,0 +1,1036 @@
+            showDRS = true;
+            listDuplicateRecordSets = getDuplicateRecordSets();
+            driFieldNames.add('DuplicateRecordSetId');
+            List<DuplicateRecordSet> duplicateRecordSets = new List<DuplicateRecordSet>((List<DuplicateRecordSet>)Database.query(constructDRSSOQL()));
```
break this line to avoid wrapping it github review window

ptewson-sfdo commented on src/classes/DRS_ContactMerge_CTRL.cls:
```diff
@@ -0,0 +1,1036 @@
+                allDRIIds.addAll(new Map<Id, DuplicateRecordItem>(dupRecSet.DuplicateRecordItems).keySet());
+                allDRIs.addAll(new Map<Id, DuplicateRecordItem>(dupRecSet.DuplicateRecordItems).values());
+            }
+            // Construct query to fetch all the related contact records to find the first contact related to the DRS record
```
break this line to avoid wrapping it github review window

ptewson-sfdo commented on src/classes/DRS_ContactMerge_CTRL.cls:
```diff
@@ -0,0 +1,1036 @@
+                Contact con = contacts.get(dri.RecordId);
+                if(!(firstContactOfDRS.containsKey(dri.DuplicateRecordSetId))) {
+                    firstContactOfDRS.put(dri.DuplicateRecordSetId,con);
+                }
```
move else to same line as close curly brace (like "} else {")

ptewson-sfdo commented on src/classes/DRS_ContactMerge_CTRL.cls:
```diff
@@ -0,0 +1,1036 @@
+    * @description Shows list of Duplicate Record Set records related to contact dupliacte rules
+    */
+    public void showDuplicateRecordSets() {
+        Try {
```
lowercase "try"

ptewson-sfdo commented on src/classes/DRS_ContactMerge_CTRL.cls:
```diff
@@ -0,0 +1,1036 @@
+            //Fetch the ids of all Duplicate Record Items
+            Set<Id> allDRIIds = new Set<Id>();
+            List<DuplicateRecordItem> allDRIs = new List<DuplicateRecordItem>();
+            for(DuplicateRecordSet dupRecSet : duplicateRecordSets) {
```
```suggestion
            for (DuplicateRecordSet dupRecSet : duplicateRecordSets) {
```

ptewson-sfdo commented on src/classes/DRS_ContactMerge_CTRL.cls:
```diff
@@ -0,0 +1,1036 @@
+                .withLimit(queryLimit)
+                .build();
+            Map<Id,Contact> contacts = new Map<Id,Contact>((List<Contact>)Database.query(contactQueryStr));
+            for(DuplicateRecordItem dri : allDRIs) {
```
```suggestion
            for (DuplicateRecordItem dri : allDRIs) {
```

ptewson-sfdo commented on src/classes/DRS_ContactMerge_CTRL.cls:
```diff
@@ -0,0 +1,1036 @@
+            Map<Id,Contact> contacts = new Map<Id,Contact>((List<Contact>)Database.query(contactQueryStr));
+            for(DuplicateRecordItem dri : allDRIs) {
+                Contact con = contacts.get(dri.RecordId);
+                if(!(firstContactOfDRS.containsKey(dri.DuplicateRecordSetId))) {
```
```suggestion
                if (!(firstContactOfDRS.containsKey(dri.DuplicateRecordSetId))) {
```

ptewson-sfdo commented on src/classes/DRS_ContactMerge_CTRL.cls:
```diff
@@ -0,0 +1,1036 @@
+                    firstContactOfDRS.put(dri.DuplicateRecordSetId,con);
+                }
+                else {
+                    if(con.createdDate < firstContactOfDRS.get(dri.DuplicateRecordSetId).createdDate) {
```
```suggestion
                    if (con.createdDate < firstContactOfDRS.get(dri.DuplicateRecordSetId).createdDate) {
```

ptewson-sfdo commented on src/classes/DRS_ContactMerge_CTRL.cls:
```diff
@@ -0,0 +1,1036 @@
+        drsFieldNames.add(subQueryStr);
+        String setOrderBy = 'LastModifiedDate DESC ';
+        String setWhereClause = 'RecordCount > 1 AND DuplicateRule.SObjectType = \'Contact\'';
+        Integer soqlLimit = 7000;
```
Why 7000? Could this be a defined constant in the class?

espl-ishita replied:
@ptewson-sfdo We have added a limit of 7000 to DRS query because we were facing the 50001 query rows issue. As we are displaying the first contact for each DRS record, we had to maintain a map with DRS id and its respective first contact record. To form this map we are querying on the duplicate contacts in the org and if the count of contact records is more we are facing the 50001 query rows issue. Hence to limit the Contacts to be retrieved we limited the associated DRS records to 7000 but as we are bulk testing we might have to lower it down to 5000. Because of this same issue we have added a limit of 30000 to the contact records query which we might have to reduce after bulk testing.

ptewson-sfdo commented on src/classes/DRS_ContactMerge_CTRL.cls:
```diff
@@ -0,0 +1,1036 @@
+                allDRIs.addAll(new Map<Id, DuplicateRecordItem>(dupRecSet.DuplicateRecordItems).values());
+            }
+            // Construct query to fetch all the related contact records to find the first contact related to the DRS record
+            Integer queryLimit = 30000;
```
Why 30000? Could this be a defined constant in the class?

ptewson-sfdo commented on src/classes/DRS_ContactMerge_CTRL.cls:
```diff
@@ -0,0 +1,1036 @@
+    */
+    public PageReference showContactRelatedToDRS() {
+        //logic to display contacts related to selected DRS
+        Try {
```
lowercase "try"

ptewson-sfdo commented on src/classes/DRS_ContactMerge_CTRL.cls:
```diff
@@ -0,0 +1,1036 @@
+    * @description This will query the duplicate contacts related to the DRS which is picked
+    * @return the list of contacts
+    */
+    public list<SObject> queryRecords() {
```
```suggestion
    public List<SObject> queryRecords() {
```

ptewson-sfdo commented on src/classes/DRS_ContactMerge_CTRL.cls:
```diff
@@ -0,0 +1,1036 @@
+
+        // Check we have atleast 2 and not more than 3 records selected for merge. If not throw an error.
+        if (mergeCandidates.size() <=1) {
+            ApexPages.addMessage(new ApexPages.Message(ApexPages.Severity.Error, Label.npe01.Contact_Merge_Error_Too_Few_Contacts));
```
break this line to avoid wrapping it github review window

ptewson-sfdo commented on src/classes/DRS_ContactMerge_CTRL.cls:
```diff
@@ -0,0 +1,1036 @@
+    private void loadMergeCandidates(set<Id> mergeCandidates) {
+
+        // Check we have atleast 2 and not more than 3 records selected for merge. If not throw an error.
+        if (mergeCandidates.size() <=1) {
```
```suggestion
        if (mergeCandidates.size() <= 1) {
```

ptewson-sfdo commented on src/classes/DRS_ContactMerge_CTRL.cls:
```diff
@@ -0,0 +1,1036 @@
+            return;
+        }
+
+        if (mergeCandidates.size() >3 ) {
```
```suggestion
        if (mergeCandidates.size() > 3) {
```

ptewson-sfdo commented on src/classes/DRS_ContactMerge_CTRL.cls:
```diff
@@ -0,0 +1,1036 @@
+        }
+
+        if (mergeCandidates.size() >3 ) {
+            ApexPages.addMessage(new ApexPages.Message(ApexPages.Severity.Error, Label.npe01.Contact_Merge_Error_Too_Many_Contacts));
```
break this line to avoid wrapping it github review window

ptewson-sfdo commented on src/classes/DRS_ContactMerge_CTRL.cls:
```diff
@@ -0,0 +1,1036 @@
+            // This is due to a bug in dynamic SOQL, quich causes querying for two real fields,
+            // e.g. npsp__foo__c and foo__c, to throw a duplicate field selected error.
+            // Only select the NPSP namespaced field and ignore the same field without namespace
+            if (!String.isBlank(UTIL_Namespace.getNamespace()) && !fieldName.startsWith(UTIL_Namespace.getNamespace()+'__') &&
```
break this line to avoid wrapping it github review window

ptewson-sfdo commented on src/classes/DRS_ContactMerge_CTRL.cls:
```diff
@@ -0,0 +1,1036 @@
+        Map<String, Schema.DescribeFieldResult> customFieldMap = new Map<String, Schema.DescribeFieldResult>();
+
+        // Construct the query string
+        String query = 'Select id, name, ';
```
```suggestion
        String query = 'SELECT Id, Name, ';
```

ptewson-sfdo commented on src/classes/DRS_ContactMerge_CTRL.cls:
```diff
@@ -0,0 +1,1036 @@
+                    //find the Name field for the related object
+                    string nameField = getRelatedObjNameField(fResult);
+
+                    if (nameField!=null) {
```
```suggestion
                    if (nameField != null) {
```

ptewson-sfdo commented on src/classes/DRS_ContactMerge_CTRL.cls:
```diff
@@ -0,0 +1,1036 @@
+        List<Contact> contacts;
+        try {
+            contacts = Database.query(query); // Query the records
+            // Ensure we got back the same number of records as expected. In case any record got deleted/moved since last search.
```
break this line to avoid wrapping it github review window

ptewson-sfdo commented on src/classes/DRS_ContactMerge_CTRL.cls:
```diff
@@ -0,0 +1,1036 @@
+            contacts = Database.query(query); // Query the records
+            // Ensure we got back the same number of records as expected. In case any record got deleted/moved since last search.
+            if (contacts == null || contacts.size() != mergeCandidates.size()) {
+                ApexPages.addMessage(new ApexPages.Message(ApexPages.Severity.Error, Label.npe01.Contact_Merge_Error_Query_Failed + ' ' + Label.npe01.Contact_Merge_Error_please_retry));
```
break this line to avoid wrapping it github review window

ptewson-sfdo commented on src/classes/DRS_ContactMerge_CTRL.cls:
```diff
@@ -0,0 +1,1036 @@
+                return;
+            }
+        } catch (Exception e) {
+            ApexPages.addMessage(new ApexPages.Message(ApexPages.Severity.Error, Label.npe01.Contact_Merge_Error_Contact_not_found + ' Error: ' + e));
```
break this line to avoid wrapping it github review window

ptewson-sfdo commented on src/classes/DRS_ContactMerge_CTRL.cls:
```diff
@@ -0,0 +1,1036 @@
+        }
+
+        if (!canUserDeleteSelectedContacts(mergeCandidates)) {
+            ApexPages.addMessage(new ApexPages.Message(ApexPages.Severity.Error, System.Label.conMergeErrorNoDeleteRecAccess));
```
break this line to avoid wrapping it github review window

ptewson-sfdo commented on src/classes/DRS_ContactMerge_CTRL.cls:
```diff
@@ -0,0 +1,1036 @@
+        for (Contact c: contacts) {
+            SObject createdBy = c.getSObject('createdby');
+            SObject modifiedBy = c.getSObject('lastModifiedBy');
+            createdByRow.values.add(new Cell(c.id, (createdBy != null ? createdBy.get('name') : '') + ' ' + c.get('createdDate')));
```
break this line to avoid wrapping it github review window

ptewson-sfdo commented on src/classes/DRS_ContactMerge_CTRL.cls:
```diff
@@ -0,0 +1,1036 @@
+            SObject createdBy = c.getSObject('createdby');
+            SObject modifiedBy = c.getSObject('lastModifiedBy');
+            createdByRow.values.add(new Cell(c.id, (createdBy != null ? createdBy.get('name') : '') + ' ' + c.get('createdDate')));
+            lastModifiedByRow.values.add(new Cell(c.id, (modifiedBy != null ? modifiedBy.get('name') : '') + ' ' + c.get('LastModifiedDate')));
```
break this line to avoid wrapping it github review window

ptewson-sfdo commented on src/classes/DRS_ContactMerge_CTRL.cls:
```diff
@@ -0,0 +1,1036 @@
+                if (row.showRadio && !row.fieldName.equals(MASTER_KEY) && row.selectedValue != master.id) {
+                    SObject selectedRecord = this.selectedRecords.get(row.selectedValue);
+
+                    // Sobject.put is not happy when sobject.get returns null. It throws an exception System.SObjectException: Illegal assignment from Object to String.
```
break this line to avoid wrapping it github review window

ptewson-sfdo commented on src/classes/DRS_ContactMerge_CTRL.cls:
```diff
@@ -0,0 +1,1036 @@
+            List<Contact> lock = [Select id from Contact where id IN :allContacts for update];
+
+            if (lock == null || lock.size() != allContacts.size()) {
+                ApexPages.addMessage(new ApexPages.Message(ApexPages.Severity.Error, label.npe01.Contact_Merge_Error_Lock_failed  + ' ' + label.npe01.Contact_Merge_Error_please_retry) );
```
break this line to avoid wrapping it github review window

ptewson-sfdo commented on src/classes/DRS_ContactMerge_CTRL.cls:
```diff
@@ -0,0 +1,1036 @@
+            } catch (Exception e) {
+                Database.rollback(sp);
+                ERR_Handler.processError(e, ERR_Handler_API.Context.CON);
+                ApexPages.addMessage(new ApexPages.Message(ApexPages.Severity.Error, label.npe01.Contact_Merge_Error_Merge_Failed + ' Error: ' + e));
```
break this line to avoid wrapping it github review window

ptewson-sfdo commented on src/classes/DRS_ContactMerge_CTRL.cls:
```diff
@@ -0,0 +1,1036 @@
+    * @description Merges the winner and losers.
+    * @return PageReference The page that it redirects to. Same page user is in.
+    */
+    public PageReference mergeContacts() {
```
This method looks identical to CON_ContactMerge_CTRL.mergeContacts(). We can't just copy the code to a new class, the original method needs to be shared here, and moved to a utility or service class if need be.

ptewson-sfdo commented on src/classes/DRS_ContactMerge_CTRL.cls:
```diff
@@ -0,0 +1,1036 @@
+    /*******************************************************************************************************
+    * @description Shows the next step of the wizard where user can see the diff of the records before merge.
+    */
+    private void loadMergeCandidates(set<Id> mergeCandidates) {
```
This method also looks identical to the variant in CON_ContactMerge_CTRL. This can't just be copy and pasted, it needs to be shared either through class inheritance or moving to a shared utility to service class.

ptewson-sfdo commented on src/classes/DRS_ContactMerge_CTRL.cls:
```diff
@@ -0,0 +1,1036 @@
+* @group-content ../../ApexDocContent/ContactMerge.htm
+*/
+
+public with sharing class DRS_ContactMerge_CTRL {
```
I'm not sold on the "DRS" prefix for this class ... I think it belongs with the "CON" prefix but maybe "CON_DRSContactMerge_CTRL". That said, I'm seeing an awful lot of duplicated code between this class and CON_ContactMerge_CTRL. It really feels like these two should share a common base class that includes the behaviors they have in common, though I haven't had the chance to dive in deeply enough to make sure that's the best way to design this. As far as I can tell there are entire methods that are copy-pasted from one class to another however, and that needs to change.

espl-ishita replied:
@ptewson-sfdo Contact merge logic implemented on the DRS Contact Merge page is same as that existing Contact Merge functionality.
We have replicated the loadMergeCandidates() method because it is declared private in CON_ContactMerge_CTRL class. Along with these we had to replicate the selectDefaultRecord(), mergeContacts() methods and its related wrapper classes as they are accessing some of the private members of CON_ContactMerge_CTRL class. We were not able to call these methods from DRS_ContactMerge_CTRL class as either they were declared private or were accessing any private members. We replicated the methods instead of creating a new base class because we wanted to make minimal changes to the already existing Contact Merge functionality code. Creating a new base class will cause a significant amount of changes to the already existing CON_ContactMerge_CTRL class. Please let us know if you want us to proceed with the new base class

espl-ishita commented:
@ptewson-sfdo  I have updated the classes as per code review comments. Two of the automatic builds are failing with the error 'Build Exception: ApexTestException'. But the tests which are causing these automatic builds to fail does not seem to be related to the updates we made in the CON_ContactMerge_CTRL class. Please have a look and let us know if you have any concerns

espl-ishita commented:
As discussed with Ali/Mike on today's(4/9/2020) call, we have updated the code to hide the 'Search Contacts' and 'See Duplicate Contact' buttons when the CON_ContactMerge page is opened from the 'Contact Merge' list button. Along with this we have added the 'Tell me more' link provided by Mike in the Custom Label
We have also cleared the the fieldRows and selectedRecords lists at the beginning of loadMergeCandidates method in the CON_ContactMerge_CTRL class so that every time only the latest selected contact to be merged will be displayed on the contact compare page.
@ptewson-sfdo  Please have a look.
